Pick a random port on the host for play kube deployments

dcermak opened the pull request:
#### Does this PR introduce a user-facing change?

```release-note
Pick a random port on the host when exposing `containerPort`s for `podman play kube`
```

`podman play kube` started to expose the port of containers in k8s deployments on the same port on the host with https://github.com/containers/podman/pull/15946. However, that breaks once replicas are involved, as they would then bind to the same port on the host.

With this PR, we change the default behavior so that podman will pick a random port for the container instead of using `containerPort`, but only if no `hostPort` is set.

As this can lead to user-unfriendly error messages, if the user sets `hostPort` and `replicas > 1`, we provide a better error message for that case.

Luap99 reviewed (changes requested):
This doesn't look right to me. I don't really know how the ports are used in k8s but ignored them here feels wrong.

I agree that the PR https://github.com/containers/podman/pull/15946 looks wrong but only because it forces the host port to equal the container port. By default we should use a random free port, e.g. compare how `podman run -p 80 ...` behaves.  This would ensure that each replica gets a unique port.

@umohnani8 WDYT?

dcermak commented:
> This doesn't look right to me. I don't really know how the ports are used in k8s but ignored them here feels wrong.
> 
> I agree that the PR #15946 looks wrong but only because it forces the host port to equal the container port. By default we should use a random free port, e.g. compare how `podman run -p 80 ...` behaves. This would ensure that each replica gets a unique port.

I am everything but a k8s expert, but according to my limited knowledge `containerPort` implies no publishing at all. The publishing & routing to the replicas is done via an ingress. So exposing all replicas on the host honestly feels wrong as well to me.

rhatdan commented:
Can we add an option to podman play kube --expose? Then we can argue which should be the default?

dcermak commented:
> Can we add an option to podman play kube --expose? Then we can argue which should be the default?

Imho setting `hostPort` is equivalent to `--expose`, so I don't really see the need for such a duplication.

rhatdan commented:
I think the difference is that when users play a kube they expect to be able to reach the ports.  Now they will get confused.

dcermak commented:
@rhatdan Done, please take a look if you like it the way it is.

rhatdan commented on pkg/specgen/generate/kube/kube.go:
```diff
@@ -25,55 +25,67 @@ import (
 )
 
-func ToPodOpt(ctx context.Context, podName string, p entities.PodCreateOptions, podYAML *v1.PodTemplateSpec) (entities.PodCreateOptions, error) {
+func ToPodOpt(ctx context.Context, podName string, p entities.PodCreateOptions, podYAML *v1apps.DeploymentSpec, exposeHostPort bool) (entities.PodCreateOptions, error) {
```
The exposeHostPort boolean should be added to the PodCreateOptions.

dcermak replied:
Done

rhatdan commented on test/e2e/play_kube_test.go:
```diff
@@ -2685,7 +2685,7 @@ spec:
 		Expect(err).ToNot(HaveOccurred())
 
-		kube := podmanTest.Podman([]string{"play", "kube", kubeYaml})
+		kube := podmanTest.Podman([]string{"play", "kube", "--expose", kubeYaml})
```
Could you add an inspect afterwards to make sure the ports are exposed?

dcermak replied:
That is already part of the test few lines below.

rhatdan commented:
Needs a rebase
@vrothberg @Luap99 @umohnani8 @haircommander PTAL

Luap99 commented:
I cannot really comment on how it should work since I do not use kubernetes.

In any case not exposing by default will cause regressions for many podman users which I consider unacceptable, they should not be forced to type --expose every time.

dcermak commented:
> In any case not exposing by default will cause regressions for many podman users which I consider unacceptable, they should not be forced to type --expose every time.

So rather change this to `--no-expose`? Would it then even make sense to have such an option?

haircommander commented:
I agree with @Luap99 that this would be breaking, but I actually think this is the correct behavior. an exposed port in kube is something the pod author opts into, and shouldn't happen by default. As far as breaking users, this may have to wait until podman has a major version bump, but this behavior looks correct to me

haircommander commented:
> I agree with @Luap99 that this would be breaking, but I actually think this is the correct behavior. an exposed port in kube is something the pod author opts into, and shouldn't happen by default. As far as breaking users, this may have to wait until podman has a major version bump, but this behavior looks correct to me

actually, wait, I have read this more clearly and I'm not sure this makes sense. I misunderstood as podman was default exposing a random port from the container to the host which could cause conflicts, not that the user specified port was conflicting.

In the [best practices](https://kubernetes.io/docs/concepts/configuration/overview/) it says
```
Don't specify a hostPort for a Pod unless it is absolutely necessary. When you bind a Pod to a hostPort, it limits the number of places the Pod can be scheduled, because each <hostIP, hostPort, protocol> combination must be unique. If you don't specify the hostIP and protocol explicitly, Kubernetes will use 0.0.0.0 as the default hostIP and TCP as the default protocol.
```
Which means maybe podman play kube should warn when a replica is being created that also has a container port? I don't think we should not expose the ports by default just because replicas will break

dcermak commented:
> actually, wait, I have read this more clearly and I'm not sure this makes sense. I misunderstood as podman was default exposing a random port from the container to the host which could cause conflicts, not that the user specified port was conflicting.

Well the current behavior is that podman will automatically expose a `containerPort` as a `hostPort` and not pick a random one instead. So replicas with a `containerPort` **cannot** work at all currently.

> Which means maybe podman play kube should warn when a replica is being created that also has a container port? I don't think we should not expose the ports by default just because replicas will break

I would suggest that we modify the behavior slightly:
- if a container defines a `containerPort` then podman will pick a random port on the host for exposing it
- if a container defines a `hostPort` as well, podman will use that one, unless replicas are involved. If replicas are involved, podman will print an error telling that this is not possible instead of the current error that you cannot bind to the specified port (which looks more like a bug)
- optionally allow to turn off port exposing via `--no-expose`?

dcermak commented:
@rhatdan @Luap99 @haircommander I have reworked this PR. The `--expose` flag is gone and podman will now pick a random port on the host unless a `hostPort` is provided. To prevent users from using replicas in that case, I have also added a new error message.

Syquel commented:
By setting the "hostPort" field automatically if not defined each and every "containerPort" is automatically published, which in my opinion is not desirable.

In Kubernetes the "containerPort" field is only used as metadata to expose / publish these ports via [Kubernetes Services](https://kubernetes.io/docs/concepts/services-networking/service/) AFAIK.  

I would like to be able to NOT publish ports on the host, which is currently not possible.  
Therefore I would like to completely remove generating the "hostPort".  
If someone wants to expose a port on the host he may set the "hostPort" field, but that choice should be up to the user and would be in line with the Kubernetes behavior.

Addendum: I have never seen "hostPort" being used in any Helm chart up until now. The Kubernetes-way to expose ports is to use [Kubernetes Services](https://kubernetes.io/docs/concepts/services-networking/service/).  
As long as Podman does not support Services the hostPort is a good workaround, but should definitely not be handled as a default.

ygalblum commented on pkg/specgen/generate/kube/kube.go:
```diff
@@ -971,10 +980,10 @@ func getPodPorts(containers []v1.Container) []types.PortMapping {
 			// only hostPort is utilized in podman context, all container ports
 			// are accessible inside the shared network namespace
-			if p.HostPort != 0 {
+			if p.HostPort != 0 && p.ContainerPort != 0 {
```
Looking at the code of the loop, this can only happen if both `ContainerPort` and `HostPort` were not set in the YAML file. I think it will be cleaner to check it in the beginning of the loop and `continue` instead of going through all the steps just to ignore them

dcermak replied:
Good catch, thanks!

ygalblum commented:
This PR conflicts in some ways with my PR #16880 that allows overriding the published ports configuration from the command line. So, I would like to resolve the discussion before either of them gets merged.

The reason I think the command line option is required is that we would like to allow users to use the the same YAML file with both K8S and Podman. Having to set the `hostPort` field in the YAML file will require users to keep two versions of the same deployment; a practice that breaks what we want to achieve.

Now, I'm not saying that this PR is wrong in creating the default ports, but since it generates the default value before where my code merges the command line with the YAML file, I would like to make sure we are not breaking each other, racing to which PR gets merged first.

dcermak commented:
> By setting the "hostPort" field automatically if not defined each and every "containerPort" is automatically published, which in my opinion is not desirable.

I have no strict opinion on this, but as this used to be the default behavior of podman so far, I was advised to modify my PR to not change the default behavior. The initial version of this PR added the `--expose` option to `podman play kube` and without that ports would not be exposed.

> In Kubernetes the "containerPort" field is only used as metadata to expose / publish these ports via [Kubernetes Services](https://kubernetes.io/docs/concepts/services-networking/service/) AFAIK.
> 
> I would like to be able to NOT publish ports on the host, which is currently not possible. Therefore I would like to completely remove generating the "hostPort". If someone wants to publish a port on the host he may set the "hostPort" field, but that choice should be up to the user and would be in line with the Kubernetes behavior.

No, that completely breaks port publishing with replicas in place (podman will create multiple copies of the same pod and try to bind every pod to the same host port, which is not possible).

> Addendum: I have never seen "hostPort" being used in any Helm chart up until now. The Kubernetes-way to expose / publish ports is to use [Kubernetes Services](https://kubernetes.io/docs/concepts/services-networking/service/). As long as Podman does not support Services the hostPort is a good workaround, but should definitely not be handled as a default.

Yes, Kubernetes uses service meshes for handling replicas, but I think that's completely out of scope for podman.

dcermak commented:
> This PR conflicts in some ways with my PR #16880 that allows overriding the published ports configuration from the command line. So, I would like to resolve the discussion before either of them gets merged.
> 
> The reason I think the command line option is required is that we would like to allow users to use the the same YAML file with both K8S and Podman. Having to set the `hostPort` field in the YAML file will require users to keep two versions of the same deployment; a practice that breaks what we want to achieve.

Well, with this PR you wouldn't have to set `hostPort` either. Podman would either use the `containerPort` or pick a random one if you are using replicas.

> 
> Now, I'm not saying that this PR is wrong in creating the default ports, but since it generates the default value before where my code merges the command line with the YAML file, I would like to make sure we are not breaking each other, racing to which PR gets merged first.

I think your PR should go in first then and I'll try to adjust this one accordingly.

ygalblum commented:
> Well, with this PR you wouldn't have to set hostPort either. Podman would either use the containerPort or pick a random one if you are using replicas.

Yes, you are right. But, it comes with some issues. Using the containerPort doesn't always work since in many cases it is a privileged port (e.g. Web Server on ports 80 or 443). Working with random port means you need to adjust the client of the endpoint on every deployment.

dcermak commented:
> > Well, with this PR you wouldn't have to set hostPort either. Podman would either use the containerPort or pick a random one if you are using replicas.
> 
> Yes, you are right. But, it comes with some issues. Using the containerPort doesn't always work since in many cases it is a privileged port (e.g. Web Server on ports 80 or 443). Working with random port means you need to adjust the client of the endpoint on every deployment.

You are of course correct. The random ports are just a workaround for the lack of a load balancer in podman.

rhatdan commented:
Isn't there another PR to help handle this situation?

rhatdan commented:
We want to get @ygalblum PR in first and then base this one off of it.

vrothberg commented:
> We want to get @ygalblum PR in first and then base this one off of it.

Do we?  I thought we gravitated more toward enforcing replicas to 1 and emit a warning because of all the limitations we face when running Podman on a single node in contrast to an orchestrated environment such as K8s.

@dcermak, how are you using replicas with Podman?  Would limiting replicas to 1 break some of your use cases?

dcermak commented:
> @dcermak, how are you using replicas with Podman? Would limiting replicas to 1 break some of your use cases?

It would break one of our openQA test cases, that was testing this specific feature (exposed ports and replicas set to 3). So limiting it to 1 would be an option.

However, we should still do something about that case to prevent users from pondering about the strange error messages that podman would produce as a result.

vrothberg commented:
> It would break one of our openQA test cases, that was testing this specific feature (exposed ports and replicas set to 3). So limiting it to 1 would be an option.

Did the test check for something particular beyond that the file can be run?

> However, we should still do something about that case to prevent users from pondering about the strange error messages that podman would produce as a result.

Do you have particular messages in mind?  I think that enforcing replicas to 1 would avoid many such errors.  Ports are one possibly more resources that are hard to "multiplex" for Podman.

dcermak commented:
> > It would break one of our openQA test cases, that was testing this specific feature (exposed ports and replicas set to 3). So limiting it to 1 would be an option.
> 
> Does the test check for something particular beyond that the file can be run?

No, it is just testing if this yaml can be deployed. Setting the replicas to 1 would be fine as well.

> > However, we should still do something about that case to prevent users from pondering about the strange error messages that podman would produce as a result.
> 
> Do you have particular messages in mind? I think that enforcing replicas to 1 would avoid many such errors. Ports are one possibly more resources that are hard to "multiplex" for Podman.

Yes, the error messages mentioned in https://github.com/containers/podman/issues/16765#issue-1481402481.

Enforcing replicas to be 1 sounds like a good idea to me. Shall I send a different PR for that instead of this one then?

vrothberg commented:
> Enforcing replicas to be 1 sounds like a good idea to me. Shall I send a different PR for that instead of this one then?

That would be great, thank you!  In case of replicas > 1, Podman should emit `logrus.Warnf` indicating to the user what's happening.  But ultimately, it shouldn't error out.

rhatdan commented:
@ygalblum PTAL

ygalblum commented:
I guess this one is now pending #17082 as well, right?

dcermak commented:
> I guess this one is now pending #17082 as well, right?

I am not entirely convinced that this PR is really needed with https://github.com/containers/podman/pull/16880 and https://github.com/containers/podman/pull/17082. At least once https://github.com/containers/podman/pull/17082 has been merged, there is really no need to handle the case of multiple replicas anymore, which is the primary issue that this PR addresses.

rhatdan commented:
Then I will close this PR.
